File: Programming_2/C++/lab_4/Rectangle.cpp
//
// Created by Danil on 17.03.2020.
//

#include "Rectangle.h"
#include "Point.h"
#include <cmath>
#include <iostream>

Rectangle::Rectangle(){
    A = Point(0, 0);
    B = Point(0, 0);
    C = Point(0, 0);
    D = Point(0, 0);
}

Rectangle::Rectangle(Point a, Point b, Point c, Point d){
    A = a;
    B = b;
    C = c;
    D = d;
}
Rectangle::Rectangle(Rectangle const &rectangle): Rectangle(rectangle.A, rectangle.B, rectangle.B, rectangle.C){
}

double Rectangle::area() const {
    return sqrt(pow(A.get_X() - B.get_X(), 2) + pow(A.get_Y() - B.get_Y(), 2))
           * sqrt(pow(B.get_X() - C.get_X(), 2) + pow(B.get_Y() - C.get_Y(), 2));
}

double Rectangle::perimeter() const {
    return (sqrt(pow(A.get_X() - B.get_X(), 2) + pow(A.get_Y() - B.get_Y(), 2))
           + sqrt(pow(B.get_X() - C.get_X(), 2) + pow(B.get_Y() - C.get_Y(), 2))) * 2;
}


std::ostream& operator<<(std::ostream&os, const Rectangle& rectangle) {
    return os  << "Points: \n A" << rectangle.A << " B" <<
                                    rectangle.B << " C" <<
                                    rectangle.C << " D" <<
                                    rectangle.D << "\nArea: " <<
                                    rectangle.area() <<"   Perimeter: " <<
                                    rectangle.perimeter() << std::endl;
}

Rectangle operator+(const Rectangle& left, const Rectangle& right) {
    return Rectangle(left.A + right.A, left.B + right.B, left.C + right.C, left.D + right.D);
}

Rectangle operator-(const Rectangle& left, const Rectangle& right) {
    return Rectangle(left.A - right.A, left.B - right.B, left.C - right.C, left.D - right.D);
}

Rectangle operator*(const Rectangle& left, const float & right) {
    return Rectangle(left.A * right, left.B * right, left.C * right, left.D * right);
}

Rectangle operator/(const Rectangle& left, const float & right) {
    return Rectangle(left.A / right, left.B / right, left.C / right, left.D / right);
}

File: Programming_2/C#/lab_4/Rectangle.cs
using System;

namespace lab_4
{
    public class Rectangle
    {
        public Point A { private get; set; }

        public Point B { get; set; }

        public Point C { get; set; }

        public Point D { get; set; }

        public Rectangle() :
            this(new Point(0, 0),
                new Point(0, 0),
                new Point(0, 0),
                new Point(0, 0))
        {
        }
        

        public Rectangle(Point a, Point b, Point c, Point d)
        {
            A = a;
            B = b;
            C = c;
            D = d;
        }
        
        // Copy constructor.
        public Rectangle(Rectangle previousRectangle) :
            this(previousRectangle.A,
                previousRectangle.B,
                previousRectangle.C,
                previousRectangle.D)
        {
        }

        public static double SideLength(Point point1, Point point2)
        {
            return Math.Sqrt(Math.Pow(point1.X - point2.X, 2) + Math.Pow(point1.Y - point2.Y, 2));
        }
        public double Area()
        {
            return SideLength(A, B) * SideLength(B, C);
        }
        
        public double Perimeter()
        {
            return (SideLength(A, B) + SideLength(B, C)) * 2;
        }
        
        //public string ShowDataRectangle => $"Point:\nA{A}\tB{B}\tC{C}\tD{D}\nArea: {Area()}\tPerimeter: {Perimeter()}";
        
        public override string ToString()
        {
            return $"Point:\nA{A}  B{B}  C{C}  D{D}\nArea: {Area()}   Perimeter: {Perimeter()}";
        }

        public static Rectangle operator +(Rectangle rectangle1, Rectangle rectangle2)
        {
            return new Rectangle(rectangle1.A + rectangle2.A,
                                rectangle1.B + rectangle2.B,
                                rectangle1.C + rectangle2.C,
                                rectangle1.D + rectangle2.D);
        }
        
        public static Rectangle operator /(Rectangle rectangle, float num)
        {
            return new Rectangle(rectangle.A / num,
                                rectangle.B / num,
                                rectangle.C / num,
                                rectangle.D / num);
        }
        
        public static Rectangle operator -(Rectangle rectangle1, Rectangle rectangle2)
        {
            return new Rectangle(rectangle1.A - rectangle2.A,
                rectangle1.B - rectangle2.B,
                rectangle1.C - rectangle2.C,
                rectangle1.D - rectangle2.D);
        }
        
        public static Rectangle operator *(Rectangle rectangle, float num)
        {
            return new Rectangle(rectangle.A * num,
                rectangle.B * num,
                rectangle.C * num,
                rectangle.D * num);
        }
    }
}